### PR TITLE
Add test for update_undocumented_columns_with_prior_knowledge function

### PR DESCRIPTION
**Pull request by syou6162**
Continued from https://github.com/z3z1ma/dbt-osmosis/pull/96.

The `update_undocumented_columns_with_prior_knowledge` function has a lot of flags (`skip_add_tags` / `skip_merge_meta` / `add_progenitor_to_meta`) involved. Therefore, it was not clear how they behave. We have added a test to this function to make it easier to follow the behavior.

In adding testing, I also corrected some behavior that differed from the documentation (the documentation says it mutates the NODE, but in fact it does not).

### CODE REVIEW

File: src/dbt_osmosis/vendored/dbt_core_interface/project.py [hunk 1]
```diff
@@ -601,9 +601,7 @@ def get_ref_node(self, target_model_name: str) -> "ManifestNode":
             ),
         )
 
-    def get_source_node(
```
Comment (syou6162) [inline src/dbt_osmosis/vendored/dbt_core_interface/project.py:604]: These changes are not related this pull request. To avoid conflict with https://github.com/z3z1ma/dbt-osmosis/pull/105, I ran linters.

File: src/dbt_osmosis/core/column_level_knowledge_propagator.py [hunk 2]
```diff
@@ -84,3 +87,51 @@ def get_node_columns_with_inherited_knowledge(
...
+            inheritables.append("tags")
+        if not skip_merge_meta:
+            inheritables.append("meta")
+
+        changes_committed = 0
+        for column in undocumented_columns:
+            prior_knowledge: ColumnLevelKnowledge = get_prior_knowledge(knowledge, column)
+            progenitor = prior_knowledge.pop("progenitor", None)
+            prior_knowledge = {k: v for k, v in prior_knowledge.items() if k in inheritables}
+            if add_progenitor_to_meta and progenitor:
```
Comment (syou6162) [inline src/dbt_osmosis/core/column_level_knowledge_propagator.py:114]: Moved forward from the original code. Because the original code mutates MODEL but did not mutate NODE.

- ref: https://github.com/z3z1ma/dbt-osmosis/blob/main/src/dbt_osmosis/core/osmosis.py#L1036-L1038

File: src/dbt_osmosis/core/column_level_knowledge_propagator.py [hunk 3]
```diff
@@ -84,3 +87,51 @@ def get_node_columns_with_inherited_knowledge(
...
+                prior_knowledge.setdefault("meta", {})
+                prior_knowledge["meta"]["osmosis_progenitor"] = progenitor
+            if not prior_knowledge:
+                continue
+            if column not in node.columns:
+                node.columns[column] = ColumnInfo.from_dict({"name": column, **prior_knowledge})
+            else:
+                node.columns[column] = ColumnInfo.from_dict(
```
Comment (syou6162) [inline src/dbt_osmosis/core/column_level_knowledge_propagator.py:122]: The original implementation was `model_column.update(prior_knowledge)`. While writing the test, I realized that this code was not updating as intended, so I fixed it. The original code had the behavior that `_extra` in `ColumnInfo` would return a modified object (the original object was not modified).

like this ↓

```python
ColumnInfo(name='first_name', description="Customer's first name. PII.", meta={}, data_type=None, constraints=[], quote=None, tags=[], _extra={'kwargs': {'name': 'first_name', 'meta': {'osmosis_progenitor': 'model.jaffle_shop_duckdb.stg_customers'}}})
```

- ref:
  - https://github.com/z3z1ma/dbt-osmosis/blob/main/src/dbt_osmosis/core/osmosis.py#L1039
  - https://github.com/dbt-labs/dbt-core/blob/d91265411061463b42d6f4d6d94e094a42a96d2b/core/dbt/contracts/graph/nodes.py#L219
  - https://github.com/dbt-labs/dbt-core/blob/d91265411061463b42d6f4d6d94e094a42a96d2b/core/dbt/contracts/util.py#L43-L45

File: src/dbt_osmosis/core/column_level_knowledge.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import re
```
Comment (syou6162) [inline src/dbt_osmosis/core/column_level_knowledge.py:1]: move from src/dbt_osmosis/core/osmosis.py to avoid cross-references with src/dbt_osmosis/core/column_level_knowledge_propagator.py.

File: tests/test_column_level_knowledge_propagator.py [hunk 5]
```diff
@@ -117,3 +120,197 @@ def test_inherit_column_level_knowledge():
     family_tree = _build_node_ancestor_tree(manifest, target_node)
     placeholders = [""]
     assert _inherit_column_level_knowledge(manifest, family_tree, placeholders) == expect
+
+
+def test_update_undocumented_columns_with_prior_knowledge():
```
Comment (syou6162) [inline tests/test_column_level_knowledge_propagator.py:125]: The code for the test added in the main section is from here on.

File: src/dbt_osmosis/core/column_level_knowledge_propagator.py (hunk 3 above)
Comment (z3z1ma) [reply src/dbt_osmosis/core/column_level_knowledge_propagator.py:122]: I see, good to know about that prop 👀

**Review by z3z1ma — APPROVED** [review]
Very nice. This is hugely beneficial for the sustainability of the software. 🎉 
Much appreciated.

**Comment by syou6162** [comment]
@z3z1ma Thank you for reviewing & I fixed conflicted codes!

File: src/dbt_osmosis/vendored/dbt_core_interface/project.py [hunk 6]
```diff
@@ -28,7 +28,6 @@
 import hmac
 import http.client as httplib
 import itertools
-import yaml
```
Comment (syou6162) [inline src/dbt_osmosis/vendored/dbt_core_interface/project.py:31]: see also https://github.com/z3z1ma/dbt-osmosis/actions/runs/6269525544/job/17026080243

**Comment by z3z1ma** [comment]
lgtm :shipit:
